ARROW-8199: [C++] Add support for multi-column sort indices on Table

kou opened the pull request:
Summary:

  * Deprecate SortToIndices()

  * Add SortIndices() because we use "sort_indices" as kernel name in
    #7240

  * Add support for sort indices in descending order on Array

  * Rename existing "sort_indices" kernel to "array_sort_indices" and
    introduce "sort_indices" meta function to support ChunkedArray,
    RecordBatch and Table

  * Require benchmark 1.5.2 or later

Benchmark:

Summary:

  * No performance regression in existing sort on array

  * Same performance as sort on array when the number of sort keys is 1

  * 1.5x-100x slower than sort on array when the number of sort keys is 2

Before:

    ----------------------------------------------------------------------------------
    Benchmark                                                     Time             CPU
    ----------------------------------------------------------------------------------
    SortToIndicesInt64Count/32768/10000/min_time:1.000        16057 ns        16054 ns
    SortToIndicesInt64Count/32768/100/min_time:1.000          16592 ns        16589 ns
    SortToIndicesInt64Count/32768/10/min_time:1.000           15979 ns        15975 ns
    SortToIndicesInt64Count/32768/2/min_time:1.000            28379 ns        28372 ns
    SortToIndicesInt64Count/32768/1/min_time:1.000             5767 ns         5766 ns
    SortToIndicesInt64Count/32768/0/min_time:1.000            12560 ns        12557 ns
    SortToIndicesInt64Count/1048576/100/min_time:1.000       729683 ns       729497 ns
    SortToIndicesInt64Count/8388608/100/min_time:1.000      6696415 ns      6693711 ns
    SortToIndicesInt64Compare/32768/10000/min_time:1.000     190488 ns       190442 ns
    SortToIndicesInt64Compare/32768/100/min_time:1.000       190879 ns       190838 ns
    SortToIndicesInt64Compare/32768/10/min_time:1.000        179156 ns       179115 ns
    SortToIndicesInt64Compare/32768/2/min_time:1.000         109098 ns       109074 ns
    SortToIndicesInt64Compare/32768/1/min_time:1.000           5682 ns         5681 ns
    SortToIndicesInt64Compare/32768/0/min_time:1.000         185022 ns       184984 ns
    SortToIndicesInt64Compare/1048576/100/min_time:1.000    9275270 ns      9273414 ns
    SortToIndicesInt64Compare/8388608/100/min_time:1.000   93126006 ns     93095276 ns

After:

    ------------------------------------------------------------------------------------------
    Benchmark                                                             Time             CPU
    ------------------------------------------------------------------------------------------
    ArraySortIndicesInt64Narrow/32768/10000/min_time:1.000            15722 ns        15721 ns
    ArraySortIndicesInt64Narrow/32768/100/min_time:1.000              16007 ns        16006 ns
    ArraySortIndicesInt64Narrow/32768/10/min_time:1.000               15178 ns        15177 ns
    ArraySortIndicesInt64Narrow/32768/2/min_time:1.000                29886 ns        29885 ns
    ArraySortIndicesInt64Narrow/32768/1/min_time:1.000                 5968 ns         5968 ns
    ArraySortIndicesInt64Narrow/32768/0/min_time:1.000                12681 ns        12681 ns
    ArraySortIndicesInt64Narrow/1048576/100/min_time:1.000           813376 ns       813331 ns
    ArraySortIndicesInt64Narrow/8388608/100/min_time:1.000          6543874 ns      6543621 ns
    ArraySortIndicesInt64Wide/32768/10000/min_time:1.000             189957 ns       189956 ns
    ArraySortIndicesInt64Wide/32768/100/min_time:1.000               190269 ns       190267 ns
    ArraySortIndicesInt64Wide/32768/10/min_time:1.000                175425 ns       175422 ns
    ArraySortIndicesInt64Wide/32768/2/min_time:1.000                 107976 ns       107975 ns
    ArraySortIndicesInt64Wide/32768/1/min_time:1.000                   5941 ns         5941 ns
    ArraySortIndicesInt64Wide/32768/0/min_time:1.000                 184858 ns       184857 ns
    ArraySortIndicesInt64Wide/1048576/100/min_time:1.000            9355194 ns      9354942 ns
    ArraySortIndicesInt64Wide/8388608/100/min_time:1.000           94101796 ns     94099551 ns
    TableSortIndicesInt64Narrow/1048576/100/16/32/min_time:1.000 1386231938 ns   1386176541 ns
    TableSortIndicesInt64Narrow/1048576/0/16/32/min_time:1.000   1350031141 ns   1349982623 ns
    TableSortIndicesInt64Narrow/1048576/100/8/32/min_time:1.000  1401741018 ns   1401632081 ns
    TableSortIndicesInt64Narrow/1048576/0/8/32/min_time:1.000    1373174328 ns   1373145968 ns
    TableSortIndicesInt64Narrow/1048576/100/2/32/min_time:1.000  1035377805 ns   1035362806 ns
    TableSortIndicesInt64Narrow/1048576/0/2/32/min_time:1.000    1035218085 ns   1035201824 ns
    TableSortIndicesInt64Narrow/1048576/100/1/32/min_time:1.000     6859319 ns      6859251 ns
    TableSortIndicesInt64Narrow/1048576/0/1/32/min_time:1.000       6847314 ns      6847048 ns
    TableSortIndicesInt64Narrow/1048576/100/16/4/min_time:1.000   626909516 ns    626904696 ns
    TableSortIndicesInt64Narrow/1048576/0/16/4/min_time:1.000     591632035 ns    591602144 ns
    TableSortIndicesInt64Narrow/1048576/100/8/4/min_time:1.000    613197155 ns    613182727 ns
    TableSortIndicesInt64Narrow/1048576/0/8/4/min_time:1.000      595568690 ns    595554829 ns
    TableSortIndicesInt64Narrow/1048576/100/2/4/min_time:1.000    424472984 ns    424453915 ns
    TableSortIndicesInt64Narrow/1048576/0/2/4/min_time:1.000      397339109 ns    397335604 ns
    TableSortIndicesInt64Narrow/1048576/100/1/4/min_time:1.000      7027310 ns      7027241 ns
    TableSortIndicesInt64Narrow/1048576/0/1/4/min_time:1.000        6891364 ns      6891272 ns
    TableSortIndicesInt64Narrow/1048576/100/16/1/min_time:1.000   516832749 ns    516823293 ns
    TableSortIndicesInt64Narrow/1048576/0/16/1/min_time:1.000     495273237 ns    495269975 ns
    TableSortIndicesInt64Narrow/1048576/100/8/1/min_time:1.000    515550360 ns    515531501 ns
    TableSortIndicesInt64Narrow/1048576/0/8/1/min_time:1.000      496670125 ns    496663084 ns
    TableSortIndicesInt64Narrow/1048576/100/2/1/min_time:1.000    340060863 ns    340053172 ns
    TableSortIndicesInt64Narrow/1048576/0/2/1/min_time:1.000      331281499 ns    331277004 ns
    TableSortIndicesInt64Narrow/1048576/100/1/1/min_time:1.000      6691062 ns      6690924 ns
    TableSortIndicesInt64Narrow/1048576/0/1/1/min_time:1.000        6384382 ns      6384323 ns
    TableSortIndicesInt64Wide/1048576/100/16/32/min_time:1.000    622544467 ns    622531557 ns
    TableSortIndicesInt64Wide/1048576/0/16/32/min_time:1.000      619193843 ns    619155597 ns
    TableSortIndicesInt64Wide/1048576/100/8/32/min_time:1.000     615885889 ns    615884764 ns
    TableSortIndicesInt64Wide/1048576/0/8/32/min_time:1.000       589917731 ns    589912005 ns
    TableSortIndicesInt64Wide/1048576/100/2/32/min_time:1.000     600951149 ns    600947256 ns
    TableSortIndicesInt64Wide/1048576/0/2/32/min_time:1.000       592304437 ns    592286953 ns
    TableSortIndicesInt64Wide/1048576/100/1/32/min_time:1.000      98781239 ns     98777123 ns
    TableSortIndicesInt64Wide/1048576/0/1/32/min_time:1.000        94136230 ns     94085863 ns
    TableSortIndicesInt64Wide/1048576/100/16/4/min_time:1.000     232323308 ns    232319347 ns
    TableSortIndicesInt64Wide/1048576/0/16/4/min_time:1.000       223708791 ns    223700834 ns
    TableSortIndicesInt64Wide/1048576/100/8/4/min_time:1.000      221975360 ns    221968035 ns
    TableSortIndicesInt64Wide/1048576/0/8/4/min_time:1.000        218214843 ns    218210306 ns
    TableSortIndicesInt64Wide/1048576/100/2/4/min_time:1.000      224756430 ns    224745751 ns
    TableSortIndicesInt64Wide/1048576/0/2/4/min_time:1.000        220809969 ns    220809044 ns
    TableSortIndicesInt64Wide/1048576/100/1/4/min_time:1.000       96159427 ns     96156899 ns
    TableSortIndicesInt64Wide/1048576/0/1/4/min_time:1.000         92307749 ns     92304526 ns
    TableSortIndicesInt64Wide/1048576/100/16/1/min_time:1.000     166390841 ns    166382427 ns
    TableSortIndicesInt64Wide/1048576/0/16/1/min_time:1.000       164576492 ns    164570581 ns
    TableSortIndicesInt64Wide/1048576/100/8/1/min_time:1.000      165724919 ns    165718584 ns
    TableSortIndicesInt64Wide/1048576/0/8/1/min_time:1.000        164048003 ns    164046222 ns
    TableSortIndicesInt64Wide/1048576/100/2/1/min_time:1.000      170131788 ns    170124950 ns
    TableSortIndicesInt64Wide/1048576/0/2/1/min_time:1.000        170874129 ns    170871245 ns
    TableSortIndicesInt64Wide/1048576/100/1/1/min_time:1.000       92314674 ns     92312953 ns
    TableSortIndicesInt64Wide/1048576/0/1/1/min_time:1.000         92023019 ns     92022643 ns

Follow-up tasks:

  * Improve performance when the number of sort keys is 2 or greater

    * Idea1: Use counting sort for small range data like on array

    * Idea2: Use threading for radix sort

  * Add support for multi-column partition Nth indices on Table

kou commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -135,5 +149,9 @@ Result<std::shared_ptr<Table>> Take(const Table& table, const ChunkedArray& indi
+Result<std::shared_ptr<Array>> SortToIndices(const Array& values, ExecContext* ctx) {
+  return SortIndices(values, ctx);
+}
+
```
This hunk is for `SortToIndices()` -> `SortIndices()` rename.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -112,11 +114,12 @@ inline void VisitRawValuesInline(const ArrayType& values,
  public:
-  void Sort(uint64_t* indices_begin, uint64_t* indices_end, const ArrayType& values) {
+  void Sort(uint64_t* indices_begin, uint64_t* indices_end, const ArrayType& values,
+            const ArraySortOptions& options) {
```
This change is for order support on array.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -125,35 +128,43 @@ class CompareSorter {
+                       [&values](uint64_t left, uint64_t right) {
+                         return values.GetView(left) > values.GetView(right);
+                       });
+    }
```
This change is for order support on array.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -112,11 +114,12 @@ inline void VisitRawValuesInline(const ArrayType& values,
 
 template <typename ArrowType>
-class CompareSorter {
+class ArrayCompareSorter {
```
`Array` prefix is added to distinguish array and table related codes.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -125,35 +128,43 @@ class CompareSorter {
+  ArrayCountSorter() = default;
 
-  explicit CountSorter(c_type min, c_type max) { SetMinMax(min, max); }
+  explicit ArrayCountSorter(c_type min, c_type max) { SetMinMax(min, max); }
```
`Array` prefix is added to distinguish array and table related codes.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -125,35 +128,43 @@ class CompareSorter {
+      SortInternal<uint32_t>(indices_begin, indices_end, values, options);
     } else {
-      SortInternal<uint64_t>(indices_begin, indices_end, values);
+      SortInternal<uint64_t>(indices_begin, indices_end, values, options);
```
This change is for order support on array.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -163,36 +174,49 @@ class CountSorter {
-        values, [&](c_type v) { indices_begin[counts[v - min_]++] = index++; },
-        [&]() { indices_begin[counts[value_range]++] = index++; });
   }
 };
```
This change is for order support on array.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -236,36 +260,40 @@ class CountOrCompareSorter {
+struct ArraySorter<Type, enable_if_t<is_floating_type<Type>::value ||
+                                     is_base_binary_type<Type>::value>> {
+  ArrayCompareSorter<Type> impl;
 };
```
`Array` prefix is added to distinguish array and table related codes.

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -299,12 +327,344 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
   }
 }
 
+class TableSorter : public TypeVisitor {
```
For multi-column sorting on table.

kou commented on cpp/src/arrow/compute/kernels/vector_sort_test.cc:
```diff
@@ -209,138 +214,155 @@ TYPED_TEST(TestNthToIndicesRandom, RandomValues) {
 
 template <typename ArrowType>
-class TestSortToIndicesKernel : public TestBase {
+class TestArraySortIndicesKernel : public TestBase {
```
This hunk is for renaming `SortToIndices()` -> `SortIndices()`. No logic change.

mrkn commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -125,35 +128,43 @@ class CompareSorter {
+    } else {
+      std::stable_sort(indices_begin, nulls_begin,
+                       [&values](uint64_t left, uint64_t right) {
+                         return values.GetView(left) > values.GetView(right);
```
This requires that all of `==`, `<`, and `>` are implemented in the value type.
Instead, writing `right < left` here only requires `==` and `<`.
IMHO, requiring only `==` and `<` is better to support extension types in the future.

kou replied:
Good catch!

kou commented:
@cyb70289 @pitrou Do you want to review this?

cyb70289 commented:
Also update kernel document? `docs/source/cpp/compute.rst`

kou commented:
Done.

cyb70289 commented:
@kou I'm okay with this patch.
As you listed in follow up tasks, sorting arrays separately and merging afterwards should be faster. And I think there are other chances to improve performance.

Some random thoughts:
- Looks you are returning a flat index array, does it make sense to return array of tuple (chunk_index, offset_in_chunk)? Maybe easier for client code to use?
- For multi column sorting, in one iteration, current code compares values column by column till first non-equal found. I don't know if a radix sort approach is better, e.g. sort by 2nd-order column first, then sort by 1st-order column. It may be possible to leverage existing array based sorting code(counting sort, etc).

kou commented:
Thanks for your comments!

> Looks you are returning a flat index array, does it make sense to return array of tuple (chunk_index, offset_in_chunk)? Maybe easier for client code to use?

Each column in table may have the different number of chunks. For example, `{"a": [[1, 2], [3], [4, 5, 6]], "b": [[1, 2, 3], [4, 5, 6]]}`. If we want to return `chunk_index` and `offset_in_chunk`, we need to return them for each column. It may decrease performance.

> For multi column sorting, in one iteration, current code compares values column by column till first non-equal found. I don't know if a radix sort approach is better, e.g. sort by 2nd-order column first, then sort by 1st-order column. It may be possible to leverage existing array based sorting code(counting sort, etc).

It's interesting! I've added the idea to follow-up tasks in the pull request description.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+    }
+
+    template <typename ArrayType>
+    ArrayType* ResolveChunk(int64_t index, int64_t& chunk_index) const {
```
Can you add a comment describing what this method does?

pitrou replied:
Also, please make it `int64_t* chunk_index`. Out-parameters should be pointers.
(you could also return a `std::pair` or a dedicated struct... anyway)

kou replied:
Added comments and changed to use new dedicated struct.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+    SortOrder order;
+    DataType* type;
+    int64_t null_count;
+    size_t num_chunks;
```
`int64_t`?

kou replied:
Fixed. It should be `int` because `ChunkedArray::num_chunks()` returns `int`.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+
+    const std::vector<ResolvedSortKey>& sort_keys() { return sort_keys_; }
+
+    bool Compare(uint64_t left, uint64_t right, size_t start_sort_key_index) {
```
Add a comment?

kou replied:
Added.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+          auto value_left = array_left->GetView(index_left);
+          auto value_right = array_right->GetView(index_right);
+          if (value_left == value_right) {
+            return comparer.Compare(left, right, 1);
```
Performance will probably be very bad if there are lots of equal values in the first column, right?

kou replied:
Confirmed. 2x-3x slower in the following cases:

The number of records: 1048576

Wide range: Almost elements will be different

Narrow range: Almost elements may be same

N sort keys   | 2           | 8           | 16
------------- | ----------- | ----------- | -----------
Narrow range  | 373997366ns | 556003297ns | 519211265ns
Wide range    | 167679014ns | 167135310ns | 166726078ns
Narrow / Wide | 2.23        | 3.32        | 3.11

In wide range cases, the number of sort keys don't affect performance.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+      array_values = values.chunk(0);
+    } else {
+      ARROW_ASSIGN_OR_RAISE(array_values,
+                            Concatenate(values.chunks(), ctx->memory_pool()));
```
This is weird. Why are you concatenating chunks for the ChunkedArray case but not for the Table case? I would expect both cases to be handled similarly (regardless of the solution).

kou replied:
Ah, I forgot to mention this.
I wanted to make ChunkedArray case as follow-up task. I wanted to focus to Table in this pull request.

But I've implemented naive ChunkedArray support that doesn't use `Concatenate()` in this pull request. It sorts each chunk and then merge them but it doesn't use threads.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+    auto out_type = uint64();
+    auto length = table.num_rows();
+    auto buffer_size = BitUtil::BytesForBits(
+        length * std::static_pointer_cast<UInt64Type>(out_type)->bit_width());
```
`sizeof(uint64_t)` perhaps?

kou replied:
If we want to use `sizeof()`, we need to use `sizeof() * 8`. Because the unit is bit not byte.
Should we use `sizeof() * 8` here?

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort_test.cc:
```diff
@@ -228,134 +233,165 @@ TYPED_TEST(TestNthToIndicesRandom, RandomValues) {
+  this->AssertSortIndices("[]", "[]");
+
+  this->AssertSortIndices("[3.4, 2.6, 6.3]", "[1, 0, 2]");
+  this->AssertSortIndices("[1.1, 2.4, 3.5, 4.3, 5.1, 6.8, 7.3]", "[0,1,2,3,4,5,6]");
```
Nit, but can you always use the same convention for spacing JSON values?

kou replied:
Added a space after `,` including existing JSON values.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort_test.cc:
```diff
@@ -364,32 +400,264 @@ TYPED_TEST(TestSortToIndicesKernelRandomCount, SortRandomValuesCount) {
+
+using RandomParam = std::tuple<std::string, double>;
+class TestTableSortIndicesRandom : public testing::TestWithParam<RandomParam> {
+  class Comparator : public TypeVisitor {
```
Can you add comments when you define non-trivial classes and methods? This would help reading the code later.

kou replied:
Added.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort_test.cc:
```diff
@@ -364,32 +400,264 @@ TYPED_TEST(TestSortToIndicesKernelRandomCount, SortRandomValuesCount) {
+#undef VISIT
+
+   private:
+    std::shared_ptr<Array> findTargetArray(std::shared_ptr<ChunkedArray> chunked_array,
```
I would expect `const Array* FindTargetArray(const ChunkedArray&, int64_t index, int64_t* chunk_index)`

kou replied:
Changed.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort_benchmark.cc:
```diff
@@ -56,17 +57,99 @@ static void SortToIndicesInt64Compare(benchmark::State& state) {
+  auto rand = random::RandomArrayGenerator(kSeed);
+
+  auto min = std::numeric_limits<int64_t>::min();
+  auto max = std::numeric_limits<int64_t>::max();
```
You will probably get almost no duplicate values with this.

kou replied:
Yes.
I've added narrow range cases that have many duplicate values.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+
+ private:
+  template <typename Type>
+  Status SortInternal() {
```
Can you add comments when you define a non-trivial class or method?

kou replied:
Added.

pitrou reviewed (commented):
I'm a bit surprised by the approach adopted here.
I would expect the following approach:
* sort(chunked array) sorts each chunk individually, then merges all of them (see `std::inplace_merge`)
* sort(table) sorts each sort column in reverse order

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -58,6 +58,34 @@ struct ARROW_EXPORT TakeOptions : public FunctionOptions {
   static TakeOptions Defaults() { return BoundsCheck(); }
 };
 
+enum SortOrder {
```
Yes. Thanks.

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -58,6 +58,34 @@ struct ARROW_EXPORT TakeOptions : public FunctionOptions {
+  explicit SortKey(std::string name, SortOrder order = ASCENDING)
+      : name(name), order(order) {}
+
+  /// The name of the sort key.
```
Thanks. Fixed.

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -58,6 +58,34 @@ struct ARROW_EXPORT TakeOptions : public FunctionOptions {
 
+enum SortOrder {
+  ASCENDING,
+  DESCENDING,
```
OK. Renamed.

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -165,8 +194,50 @@ Result<std::shared_ptr<Array>> NthToIndices(const Array& values, int64_t n,
+/// \param[in] ctx the function execution context, optional
+/// \return offsets indices that would sort an array
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> SortIndices(const Array& values, SortOrder order,
```
I wanted to accept `SortIndices(array, ctx)` usage but it may not be used.
I've removed the SortIndices(array, ctx)` and added `SortOrder::Ascending` as the default value.

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -165,8 +194,50 @@ Result<std::shared_ptr<Array>> NthToIndices(const Array& values, int64_t n,
+///
+/// Perform an indirect sort of array. The output array will contain
+/// indices that would sort an array, which would be the same length
+/// as input. Nulls will be stably partitioned to the end of the output.
```
Added.

kou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -58,6 +58,34 @@ struct ARROW_EXPORT TakeOptions : public FunctionOptions {
+  SortOrder order;
+};
+
+struct ARROW_EXPORT SortOptions : public FunctionOptions {
```
I want to keep this because this is used by all inputs (`Array`, `ChunkedArray`, `RecordBatch` and `Table`) with the `sort_indices()` compute kernel.
So I think that no `Table` prefix is suitable.

kou commented:
>     * sort(chunked array) sorts each chunk individually, then merges all of them (see `std::inplace_merge`)

I've implemented.
See also: https://github.com/apache/arrow/pull/8612#discussion_r526616431

>     * sort(table) sorts each sort column in reverse order

I wanted this a follow-up task but I've implemented this in this pull request. Because both of them mentions this approach.

My implementation `TableRadixSorter` is naive but this approach is slower than the first sort key based approach:

Summary:

  * The first sort key approach is 1.1x-27.2x faster than the radix sort approach except narrow range and 2 sort keys case.

  * The first sort key approach doesn't depend on the number of sort keys in wide range data.

  * The first sort key approach slowed down when the number of sort keys is increased in narrow range data.

  * The radix sort approach always slowed down when the number of sort keys is increased.
    * Because the radix sort approach always sorts all sort keys.

----

Description:

N records: 1048576

Narrow range: Almost elements may be same

Wide range: Almost elements will be different

----

N chunks: 4

Narrow range:

N sort keys          | 2           | 8            | 16
-------------------- | ----------- | ------------ | ------------
Radix                | 318651531ns | 2353793768ns | 4627874931ns
First sort key       | 397339109ns | 595568690ns  | 591632035ns
Radix/First sort key | 0.8         | 3.9          | 7.8

Wide range:

N sort keys          | 2           | 8            | 16
-------------------- | ----------- | ------------ | ------------
Radix                | 593904971ns | 2989364923ns | 6101922415ns
First sort key       | 220809969ns | 218214843ns  | 223708791ns
Radix/First sort key | 2.6         | 13.6         | 27.2

----

N chunks: 32

Narrow range:

N sort keys          | 2            | 8            | 16
-------------------- | ------------ | ------------ | ------------
Radix                | 1198835057ns | 6513599980ns | 13562068073ns
First sort key       | 1035218085ns | 1373174328ns | 1350031141ns
Radix/First sort key | 1.1          | 4.7          | 10.0

Wide range:

N sort keys          | 2            | 8            | 16
-------------------- | ------------ | ------------ | ------------
Radix                | 1585765386ns | 7305830182ns | 15323045709ns
First sort key       | 592304437ns  | 589917731ns  | 619193843ns
Radix/First sort key | 2.6          | 12.3         | 24.7

----

Can we mark the radix sort approach improvement a follow-up task? Or should we optimize the radix sort approach in this pull request?

kou commented on cpp/src/arrow/compute/kernels/vector_sort.cc:
```diff
@@ -346,14 +499,644 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+};
+
+// Sort a table from the beginning.
+class TableFromTheBeginningSorter : public TypeVisitor {
```
Can anyone suggest better name...?

kou commented:
@pitrou Could you review this again?

pitrou commented:
Thanks for the numbers. I didn't expect "first sort key" to be better, but I'm convinced now.

pitrou reviewed (approved):
+1. Thanks a lot @kou, very nice work. I've pushed some updates and will merge.
